dl1_data_handler/processor.py

```diff
@@ -0,0 +1,73 @@
+import copy
+
+import numpy as np
+
+class DL1DataProcessor():
+
+    def __init__(self, mode, input_description, transforms=None):
+        if mode in ['mono', 'stereo', 'multi-stereo']:
+            self.mode = mode
+        else:
+            raise ValueError("Invalid mode selection '{}'. Valid options: "
+                             "'mono', 'stereo', 'multi-stereo'".format(mode))
+        if transforms is None:
+            transforms = []
+        self.transforms = transforms
+        self.input_description = copy.deepcopy(input_description)
+        for transform in self.transforms:
+            input_description = transform.describe(input_description)
+        self.output_description = input_description
+
+    def process(self, example):
+        for transform in self.transforms:
+            example = transform.transform(example)
+        return example
+
+class Transform():
+
+    def __init__(self):
+        self.description = []
+
+    def describe(self, description):
+        self.description = description
+        return self.description
+
+    def transform(self, example):
+        return example
+
+class ConvertParticleIDToClassLabel(Transform):
+
+    def __init__(self):
+        super().__init__()
+        self.particle_id_to_class = {
+            0: 1, # gamma
+            101: 0 # proton
+            }
+
+    def describe(self, description):
+        self.description = [
+            {**des, 'name': 'class_label'} for des
+            in description if des['name'] == 'particle_id']
+        return self.description
+
+    def transform(self, example):
```

Comment:
Should transform() somehow implement a check that the example actually matches the description? It seems to me that in theory you could have an example with the same length as self.description but with the wrong order or different fields. The way things are set up now that wouldn't occur, but I feel like the Transform class is such a nice and useful design that I could imagine it being used in other settings, where it might be nice for each Transform to be more self-contained/robust.

Reply:
This is a good idea. Maybe a good way to do this would be to add a validate option to the transform method that checks that the transformed example matches the description in terms of the number of elements and their types and shapes.

Reply:
the Transform class indeed looks great. To go a bit further it could be a callable:
`class Transform():
    def __init__(self):
        self.description = []
    def describe(self, description):
        self.description = description
        return self.description
    def __call__(self, example):
        return example`
turning the process function of DL1DataProcessor into:
`def process(self, example):
        for transform in self.transforms:
            example = transform(example)
        return example`

Reply:
I turned Transform into a callable in b960247 and added a validation method in ce95dbc. Note that validation is performed at processing, not initialization, time. This is unavoidable because validation requires comparison with a specific example. I envision the workflow being turning on validation while testing the network and turning it off for final production to optimize speed (though the difference is likely small).